sklearn/model_selection/_validation.py

```diff
@@ -732,12 +732,13 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
     predictions = func(X_test)
     if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
         n_classes = len(set(y))
-        predictions_ = np.zeros((_num_samples(X_test), n_classes))
-        if method == 'decision_function' and len(estimator.classes_) == 2:
-            predictions_[:, estimator.classes_[-1]] = predictions
-        else:
-            predictions_[:, estimator.classes_] = predictions
-        predictions = predictions_
+        if n_classes != len(estimator.classes_):
```

Comment:
Should we raise a warning in this case?

Reply:
This implicitly checks whether ``n_classes > 2``, right? If there was an estimator that would work for 1 class, the bug would still appear, right? Should we add an explicit check that ``n_classes > 2`` to guard against that?

Reply:
I don't see where the ``ovo`` case is addressed. We should also raise an error if ``predictions.shape[1] != len(estimator.n_classes_)``, right?

The test for this case is in ``test_cross_val_predict_with_method``. I suggest we add a test with ``SVC(decision_function_shape='ovo')`` and make sure a helpful error is thrown.

Reply:
@amueller I agree that a warning should be raised in this case.

I don't follow your statement about `n_classes > 2`. Which bug are you referring to? Could you give a specific example?

For your third point, what I've fixed here is the following regression of `cross_val_predict` behavior when `decision_function_shape=ovo`

Code to test:
```python
from sklearn.datasets import load_digits
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_predict

X, y = load_digits(return_X_y=True)
cross_val_predict(SVC(kernel='linear', decision_function_shape='ovo'), X, y, method='decision_function').shape
```

**Expected Results**
In previous versions, this is `(1797, 45)`

**Actual Results**
```
ValueError: shape mismatch: value array of shape (602,45) could not be broadcast to indexing result of shape (10,602)
```

I've also included the test for the `ovo` case in Line 792-798 in `test_validation`. Should I move this somewhere else?

I also acknowledge that the `ovo` case still throws an error if the folds are not properly stratified. However, this has always been the case even in previous versions, and is due to the special shape of `decision_function` for `ovo`.

Reply:
> I also acknowledge that the ovo case still throws an error if the folds are not properly stratified. However, this has always been the case even in previous versions, and is due to the weird shape of decision_function for ovo.

That was the case I meant. I just wanted a better error in this case, and I want a test for that error.